File: backend_python/app/repositories/firestore_repository.py
"""
Generic Firestore repository - mirrors IFirestoreRepository
"""
from google.cloud import firestore
from typing import TypeVar, Generic, Optional, Any
import logging

logger = logging.getLogger(__name__)

T = TypeVar('T')


class FirestoreRepository:
    """
    Generic Firestore repository for document operations.
    Mirrors the .NET IFirestoreRepository interface.
    """
    
    def __init__(self, db: firestore.Client):
        self.db = db
    
    async def get_document(self, collection: str, document_id: str) -> Optional[dict[str, Any]]:
        """
        Get a single document by ID.
        Mirrors GetDocumentAsync<T>.
        """
        try:
            doc_ref = self.db.collection(collection).document(document_id)
            doc_snapshot = doc_ref.get()
            
            if not doc_snapshot.exists:
                return None
            
            data = doc_snapshot.to_dict()
            if data:
                data['id'] = doc_snapshot.id
            return data
        except Exception as e:
            logger.error(f"Error getting document {document_id} from {collection}: {e}")
            raise
    
    async def create_document(
        self, 
        collection: str, 
        document: dict[str, Any], 
        document_id: Optional[str] = None
    ) -> str:
        """
        Create a document in a collection.
        Mirrors CreateDocumentAsync<T>.
        """
        try:
            collection_ref = self.db.collection(collection)
            
            # Remove 'id' from document if present (Firestore manages IDs)
            doc_data = {k: v for k, v in document.items() if k != 'id'}
            
            if document_id:
                doc_ref = collection_ref.document(document_id)
                doc_ref.set(doc_data)
                return document_id
            else:
                # Auto-generate ID
                doc_ref = collection_ref.document()
                doc_ref.set(doc_data)
                return doc_ref.id
        except Exception as e:
            logger.error(f"Error creating document in {collection}: {e}")
            raise
    
    async def update_document(
        self, 
        collection: str, 
        document_id: str, 
        document: dict[str, Any]
    ) -> None:
        """
        Update a document (merge with existing).
        Mirrors UpdateDocumentAsync<T>.
        """
        try:
            doc_ref = self.db.collection(collection).document(document_id)
            # Remove 'id' from document if present
            doc_data = {k: v for k, v in document.items() if k != 'id'}
            doc_ref.set(doc_data, merge=True)
        except Exception as e:
            logger.error(f"Error updating document {document_id} in {collection}: {e}")
            raise
    
    async def delete_document(self, collection: str, document_id: str) -> None:
        """
        Delete a document.
        Mirrors DeleteDocumentAsync.
        """
        try:
            doc_ref = self.db.collection(collection).document(document_id)
            doc_ref.delete()
        except Exception as e:
            logger.error(f"Error deleting document {document_id} from {collection}: {e}")
            raise
    
    async def get_documents(
        self, 
        collection: str, 
        field: Optional[str] = None, 
        value: Optional[Any] = None
    ) -> list[dict[str, Any]]:
        """
        Get multiple documents, optionally filtered by field.
        Mirrors GetDocumentsAsync<T>.
        """
        try:
            query = self.db.collection(collection)
            
            if field and value is not None:
                query = query.where(field, "==", value)
            
            docs = query.stream()
            results = []
            for doc in docs:
                data = doc.to_dict()
                if data:
                    data['id'] = doc.id
                results.append(data)
            
            return results
        except Exception as e:
            logger.error(f"Error getting documents from {collection}: {e}")
            raise

File: backend/CricketTournament.Infrastructure/Repositories/FirestoreRepository.cs
using CricketTournament.Domain.Interfaces;
using Google.Cloud.Firestore;
using FirebaseAdmin;

namespace CricketTournament.Infrastructure.Repositories;

public class FirestoreRepository : IFirestoreRepository
{
    private readonly FirestoreDb _db;

    public FirestoreRepository()
    {
        _db = FirestoreDb.Create(FirebaseAdmin.FirebaseApp.DefaultInstance.ProjectId);
    }

    public async Task<T?> GetDocumentAsync<T>(string collection, string documentId) where T : class
    {
        var docRef = _db.Collection(collection).Document(documentId);
        var snapshot = await docRef.GetSnapshotAsync();
        
        if (!snapshot.Exists) return null;
        
        return snapshot.ConvertTo<T>();
    }

    public async Task<string> CreateDocumentAsync<T>(string collection, T document, string? documentId = null) where T : class
    {
        var collectionRef = _db.Collection(collection);
        DocumentReference docRef;

        if (string.IsNullOrEmpty(documentId))
        {
            docRef = await collectionRef.AddAsync(document);
        }
        else
        {
            docRef = collectionRef.Document(documentId);
            await docRef.SetAsync(document);
        }

        return docRef.Id;
    }

    public async Task UpdateDocumentAsync<T>(string collection, string documentId, T document) where T : class
    {
        var docRef = _db.Collection(collection).Document(documentId);
        await docRef.SetAsync(document, SetOptions.MergeAll);
    }

    public async Task DeleteDocumentAsync(string collection, string documentId)
    {
        var docRef = _db.Collection(collection).Document(documentId);
        await docRef.DeleteAsync();
    }

    public async Task<List<T>> GetDocumentsAsync<T>(string collection, string? field = null, object? value = null) where T : class
    {
        var query = _db.Collection(collection);

        if (!string.IsNullOrEmpty(field) && value != null)
        {
            query = query.WhereEqualTo(field, value);
        }

        var snapshot = await query.GetSnapshotAsync();
        return snapshot.Documents.Select(doc => doc.ConvertTo<T>()).ToList();
    }
}

